tests/haskell/Network/WebSockets/Routing/Tests.hs

```diff
@@ -0,0 +1,228 @@
+        , do noTrailingSlash
+             routeAccept $ sendTextData `flip` ("No trailing" :: Text)
+        ]
+
+    expectTrailing con = do
+        msg <- receiveData con
+        return $ msg == ("Trailing" :: Text)
+
+    expectNoTrailing con = do
+        msg <- receiveData con
+        return $ msg == ("No trailing" :: Text)
+
+
+--------------------------------------------------------------------------------
+testPathRoutes :: Assertion
+testPathRoutes = withServer port server $ do
+
+    assert =<< runClient "127.0.0.1" port "/path/foo"   (expect "foo")
+    assert =<< runClient "127.0.0.1" port "/path/bar"   (expect "bar")
+    assert =<< runClient "127.0.0.1" port "/path/"      (expect "none")
+
+  where
+    port = 42952
+
+    server = routeWebSockets $ dir "path" $ msum
+        [ path $ \p -> do
+            routeAccept $ sendTextData `flip` (pack p)
+        , routeAccept $ sendTextData `flip` ("none" :: Text)
+        ]
+
+    expect expected con = do
+        msg <- receiveData con
+        return $ (msg :: Text) == expected
+
+
+--------------------------------------------------------------------------------
+testSubprotocolRoutes :: Assertion
+testSubprotocolRoutes = withServer port server $ do
+
+    assert =<< runClientWith "127.0.0.1" port "/" opts (proto "foo") (expect "foo")
+    assert =<< runClientWith "127.0.0.1" port "/" opts (proto "bar") (expect "bar")
+    assert =<< runClient     "127.0.0.1" port "/"                    (expect "none")
+
+  where
+    port = 42953
+    opts = defaultConnectionOptions
+
+    -- Header with subprotocol definition
+    proto p = [("Sec-WebSocket-Protocol", p)]
+
+    server = routeWebSockets $ msum
+        [ subprotocol "foo" $ routeAccept $ sendTextData `flip` ("foo" :: Text)
+        , subprotocol "bar" $ routeAccept $ sendTextData `flip` ("bar" :: Text)
+        , do
+            noSubprotocols
+            routeAccept $ sendTextData `flip` ("none" :: Text)
+        ]
+
+    expect expected con = do
+        msg <- receiveData con
+        return $ (msg :: Text) == expected
+
+--------------------------------------------------------------------------------
+testQueryRoutes :: Assertion
+testQueryRoutes = withServer port server $ do
+
+    assert =<< runClient "127.0.0.1" port "/foo?q=a"            (expect "foo")
+    assert =<< runClient "127.0.0.1" port "/foo/bar?q=a"        (expect "bar")
+    assert =<< runClient "127.0.0.1" port "/foo/bar2?q=a"       (expect "bar2")
+    assert =<< runClient "127.0.0.1" port "/path/foo?q=a"       (expect "foo")
+    assert =<< runClient "127.0.0.1" port "/path/?q=a"          (expect "none")
+    assert =<< runClient "127.0.0.1" port "/trailing/?q=a"      (expect "trailing")
+    assert =<< runClient "127.0.0.1" port "/trailing?q=a"       (expect "no trailing")
+    assert =<< runClient "127.0.0.1" port "/full?path=reply"    (expect "/full?path=reply")
+    assert =<< runClient "127.0.0.1" port "/full/2?path=reply"  (expect "/2?path=reply")
```

Comment:
If it is possible to have `/` at the right of `?`, maybe we should add a test with a path like `"/foo/bar?q=ab/cd/e"`